crypto/signatures.go

```diff
@@ -23,25 +23,58 @@ type (
 var (
 	ErrNilInput         = errors.New("cannot use nil input")
 	ErrInvalidSignature = errors.New("invalid signature")
+	ErrRandUnexpected   = errors.New("unexpected result from random number generator")
 )
 
-// GenerateKeyPair creates a public-secret keypair that can be used to sign and
-// verify messages.
-func GenerateSignatureKeys() (sk SecretKey, pk PublicKey, err error) {
+// KeyPairGenerator is an interface that allows the caller to generate
+// public-secret key pairs.
+type KeyPairGenerator interface {
+	Generate() (SecretKey, PublicKey, error)
+	GenerateDetermistic(entropy [EntropySize]byte) (SecretKey, PublicKey)
+}
+
+// readBytesFunc is a function pointer that reads bytes into a buffer and
+// returns the total number of bytes written to the buffer.
+type readBytesFunc func([]byte) (int, error)
```

Comment:
I don't think that works because io.Reader is an interface that contains a function that matches this signature, but we need a function pointer for the signature. I tried changing the type of `SignatureKeyGenerator.readRandBytes` and I get:

```
crypto/signatures.go:54: cannot use "crypto/rand".Read (type func([]byte) (int, error)) as type io.Reader in field value:
        func([]byte) (int, error) does not implement io.Reader (missing Read method)
crypto/signatures.go:61: cannot call non-function skg.readRandBytes (type io.Reader)
crypto/signatures_test.go:53: cannot use mockRandGenerator.Read (type func([]byte) (int, error)) as type io.Reader in field value:
        func([]byte) (int, error) does not implement io.Reader (missing Read method)
crypto/signatures_test.go:78: cannot use mockRandGenerator.Read (type func([]byte) (int, error)) as type io.Reader in field value:
        func([]byte) (int, error) does not implement io.Reader (missing Read method)
crypto/signatures_test.go:96: cannot use mockRandGenerator.Read (type func([]byte) (int, error)) as type io.Reader in field value:
        func([]byte) (int, error) does not implement io.Reader (missing Read method)
```